File: web/src/render/footnoteListRenderer.ts
import type MarkdownNode from "../types/MarkdownNode";
import type Renderer from "../types/Renderer";
import type RendererState from "../types/RendererState";
import renderChildren from "./renderChildren";

const renderer: Renderer = {
	name: "footnote_list",
	render,
};
export default renderer;

function render(_node: MarkdownNode, state: RendererState): void {
	state.output += `<section class="footnotes">\n<ol>\n`;
	let number = 1;
	for (let node of state.footnotes) {
		let label = number++;
		let id = `fn${label}`;
		let href = `#fnref${label}`;
		state.output += `<li id="${id}">`;
		let refNode = state.footnoteRefs[node.info!];
		if (refNode !== undefined) {
			renderChildren(refNode, state);
		}
		if (state.output.endsWith("</p>\n")) {
			state.output = state.output.slice(0, state.output.length - 5);
		}
		state.output += ` <a href="${href}" class="footnote-backref">↩</a></p>\n</li>\n`;
	}
	state.output += `</ol>\n</section>`;
}

File: zig/src/render/footnoteListRenderer.zig
const std = @import("std");
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const RendererState = @import("../types/RendererState.zig").RendererState;
const Renderer = @import("../types/Renderer.zig").Renderer;
const renderChildrenFn = @import("renderChildren.zig").renderChildren;

pub const footnoteListRenderer = Renderer{
    .name = "footnote_list",
    .render = render,
};

pub fn renderFootnoteList(state: *RendererState) void {
    render(null, state, null, null, null);
}

pub fn render(_node: ?*const MarkdownNode, state: *RendererState, decode: ?bool) void {
    _ = _node;
    _ = decode;
    state.output.appendSlice(state.allocator, "<section class=\"footnotes\">\n<ol>\n") catch unreachable;

    var number: usize = 1;
    for (state.footnotes.items) |node| {
        const label = number;
        number += 1;

        const id = std.fmt.allocPrint(state.allocator, "fn{d}", .{label}) catch unreachable;
        defer state.allocator.free(id);
        const href = std.fmt.allocPrint(state.allocator, "#fnref{d}", .{label}) catch unreachable;
        defer state.allocator.free(href);

        const open_tag = std.fmt.allocPrint(state.allocator, "<li id=\"{s}\">", .{id}) catch unreachable;
        defer state.allocator.free(open_tag);
        state.output.appendSlice(state.allocator, open_tag) catch unreachable;

        renderChildrenFn(node, state, true);

        if (state.output.items.len >= 5 and std.mem.eql(u8, state.output.items[state.output.items.len - 5 ..], "</p>\n")) {
            const new_len = state.output.items.len - 5;
            state.output.shrinkRetainingCapacity(new_len);
        }

        const close_tag = std.fmt.allocPrint(state.allocator, " <a href=\"{s}\" class=\"footnote-backref\">↩</a></p>\n</li>\n", .{href}) catch unreachable;
        defer state.allocator.free(close_tag);
        state.output.appendSlice(state.allocator, close_tag) catch unreachable;
    }

    state.output.appendSlice(state.allocator, "</ol>\n</section>") catch unreachable;
}
